statsd/receiver.go

```diff
@@ -145,21 +143,34 @@ func (mr *metricReceiver) handleMessage(ctx context.Context, addr net.Addr, msg
 				triedToGetTags = true
 				additionalTags = mr.getAdditionalTags(addr.String())
 			}
-			if len(additionalTags) > 0 {
+			if metric != nil {
+				numMetrics++
+				metric.Tags = append(metric.Tags, mr.tags...)
 				metric.Tags = append(metric.Tags, additionalTags...)
+				err = mr.handler.DispatchMetric(ctx, metric)
+			} else {
+				numEvents++
+				event.Tags = append(event.Tags, mr.tags...)
+				event.Tags = append(event.Tags, additionalTags...)
+				if event.DateHappened == 0 {
+					event.DateHappened = time.Now().Unix()
+				}
+				err = mr.handler.DispatchEvent(ctx, event)
 			}
-			if err := mr.handler.HandleMetric(ctx, metric); err != nil {
-				return err
+			if err != nil {
+				exitError = err
+				break
 			}
-			numMetrics++
 		}
 
 		if readerr == io.EOF {
 			// if was EOF, finished handling
-			atomic.AddUint64(&mr.metricsReceived, uint64(numMetrics))
-			return nil
+			break
 		}
 	}
+	atomic.AddUint64(&mr.metricsReceived, uint64(numMetrics))
+	atomic.AddUint64(&mr.eventsReceived, uint64(numEvents))
```

Comment:
This change is to make sure we increment counters even in case of early exit (i.e. due to error).